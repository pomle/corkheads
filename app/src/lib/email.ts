export function isEmailValid(email: string) {
  return /^\S+@\S+\.\S+$/.test(email);
}
