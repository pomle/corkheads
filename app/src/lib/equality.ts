export function listEquals(a: any[], b: any[]): boolean {
  if (a.length !== b.length) {
    return false;
  }

  for (let i = 0; i < a.length; i += 1) {
    if (a[i] !== b[i]) {
      return false;
    }
  }

  return true;
}
