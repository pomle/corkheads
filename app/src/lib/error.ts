export function asSilentError(error: Error) {
  console.error(error); // eslint-disable-line no-console
}
