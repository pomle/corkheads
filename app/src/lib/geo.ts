export function clonePosition(position: Position): Position {
  const { coords } = position;
  return {
    coords: {
      latitude: coords.latitude,
      longitude: coords.longitude,
      altitude: coords.altitude,
      accuracy: coords.accuracy,
      altitudeAccuracy: coords.altitudeAccuracy,
      heading: coords.heading,
      speed: coords.speed,
    },
    timestamp: position.timestamp,
  };
}
