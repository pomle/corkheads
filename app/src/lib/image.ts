export type Size = {
  width: number;
  height: number;
};

type Format = "image/jpeg" | "image/png" | "image/webp";

export function loadImage(url: string) {
  return new Promise<HTMLImageElement>((resolve, reject) => {
    const image = new Image();
    image.addEventListener("load", () => resolve(image));
    image.addEventListener("error", reject);
    image.src = url;
  });
}

export function resize(
  image: HTMLImageElement | HTMLCanvasElement,
  size: Size
): HTMLCanvasElement {
  const canvas = document.createElement("canvas");
  canvas.width = size.width;
  canvas.height = size.height;
  const context = canvas.getContext("2d");
  if (!context) {
    throw new Error("Could not get context from canvas");
  }
  context.drawImage(image, 0, 0, canvas.width, canvas.height);
  return canvas;
}

export function createBlob(
  canvas: HTMLCanvasElement,
  format: Format,
  quality?: number
) {
  return new Promise<Blob>((resolve, reject) => {
    try {
      canvas.toBlob(
        (blob) => {
          if (!blob) {
            reject(new Error("Blob object missing"));
            return;
          }

          resolve(blob);
        },
        format,
        quality
      );
    } catch (error) {
      reject(error);
    }
  });
}
