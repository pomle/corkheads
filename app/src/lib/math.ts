export class Vec {
  x: number;

  y: number;

  constructor(x: number, y: number) {
    this.x = x;
    this.y = y;
  }

  get length() {
    return Math.sqrt(this.x * this.x + this.y * this.y);
  }

  set length(length) {
    this.normalize();
    this.x *= length;
    this.y *= length;
  }

  add(vec: Vec) {
    this.x += vec.x;
    this.y += vec.y;
  }

  subtract(vec: Vec) {
    this.x -= vec.x;
    this.y -= vec.y;
  }

  normalize() {
    const length = this.length;
    this.x /= length;
    this.y /= length;
  }
}

export const clamp = (value: number, min: number, max: number) => {
  if (value < min) {
    return min;
  }
  if (value > max) {
    return max;
  }
  return value;
};

type Rect = {
  x: number;
  y: number;
};

export function fitRect(rect: Rect, bounds: Rect): Rect {
  const aspectRatio = rect.x / rect.y;

  const size = {
    x: bounds.y * aspectRatio,
    y: bounds.x / aspectRatio,
  };

  if (size.y > bounds.y) {
    size.y = bounds.y;
  } else {
    size.x = bounds.x;
  }

  return { ...size };
}

export function shrinkToFitRect(rect: Rect, bounds: Rect): Rect {
  if (rect.x < bounds.x && rect.y < bounds.y) {
    return { ...rect };
  }

  return fitRect(rect, bounds);
}
