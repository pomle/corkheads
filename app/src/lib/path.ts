import { useMemo } from "react";
import { useHistory, useParams } from "react-router-dom";

type ParamCodec<SourceType> = {
  encode: (source: SourceType) => string;
  decode: (param: string) => SourceType;
};

export type PathCodec = {
  [key: string]: ParamCodec<any>;
};

type Params<Source extends {}> = {
  [key in keyof Source]: string;
};

type InputParams<CodecType extends PathCodec> = {
  [Prop in keyof CodecType]: Parameters<CodecType[Prop]["decode"]>[0];
};

type OutputParams<CodecType extends PathCodec> = {
  [Prop in keyof CodecType]: ReturnType<CodecType[Prop]["decode"]>;
};

export interface Path<PathCodecType extends PathCodec> {
  path: string;
  codec: PathCodecType;
  url(params: InputParams<PathCodecType>): string;
  encode(params: InputParams<PathCodecType>): Params<PathCodecType>;
  decode(params: Params<PathCodecType>): OutputParams<PathCodecType>;
  append<AdditionalPathCodecType extends PathCodec>(
    pathName: string,
    codec: AdditionalPathCodecType
  ): Path<PathCodecType & AdditionalPathCodecType>;
}

export function createPath<PathCodecType extends PathCodec>(
  pathName: string,
  codec: PathCodecType
): Path<PathCodecType> {
  return {
    path: pathName,
    codec,

    encode(params) {
      const encoded: any = {};
      for (const key of Object.keys(params)) {
        encoded[key] = codec[key].encode(params[key]);
      }
      return encoded;
    },

    decode(params) {
      const decoded: any = {};
      for (const key of Object.keys(params)) {
        decoded[key] = codec[key].decode(params[key]);
      }
      return decoded;
    },

    url(params) {
      const encoded = this.encode(params);
      return Object.entries(encoded).reduce((pathName, [key, value]) => {
        return pathName.replace(":" + key, value);
      }, pathName);
    },

    append<AdditionalPathCodecType extends PathCodec>(
      pathName: string,
      addCodec: AdditionalPathCodecType
    ) {
      return createPath(this.path + pathName, { ...codec, ...addCodec });
    },
  };
}

export function createCodec<T>(
  encode: (source: T) => string,
  decode: (source: string) => T
): ParamCodec<T> {
  return {
    encode,
    decode,
  };
}

export function usePath<PathCodecType extends PathCodec>(
  path: Path<PathCodecType>
) {
  const raw = useParams<Params<PathCodecType>>();
  const params = path.decode(raw);

  return {
    params,
    url() {
      return path.url(params);
    },
  };
}

type PathSet = Record<string, Path<PathCodec>>;

type PathNavigator<T extends Path<PathCodec>> = (
  params: Parameters<T["url"]>[0]
) => void;

type PathNavigatorSet<T extends PathSet> = {
  [K in keyof T]: PathNavigator<T[K]>;
};

export function useNavigator<T extends PathSet>(paths: T): PathNavigatorSet<T> {
  const history = useHistory();

  return useMemo(() => {
    return Object.entries(paths).reduce((routes, [key, path]) => {
      routes[key] = (params: any) => {
        const url = path.url(params);
        history.push(url);
      };

      return routes;
    }, {} as Record<string, any>) as PathNavigatorSet<T>;
  }, [history, paths]);
}
