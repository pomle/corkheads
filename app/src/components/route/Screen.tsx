import React, { useEffect, useRef, useState } from "react";
import { useRouteMatch } from "react-router-dom";
import { PathCodec, Path } from "lib/path";

const Screen = <PathCodecType extends PathCodec>({
  path,
  exact = false,
  unmount = 5000,
  transition: Transition,
  children: Component,
}: {
  path: Path<PathCodecType>;
  exact?: boolean;
  unmount?: number;
  transition?: React.FC<{ active: boolean }>;
  children: (
    params: ReturnType<Path<PathCodecType>["decode"]>
  ) => React.ReactElement;
}) => {
  const [mount, setMount] = useState<boolean>(true);

  const element = useRef<React.ReactElement | null>(null);
  const match = useRouteMatch<any>(path.path);

  const active = exact ? !!(match && match.isExact) : !!match;

  if (match) {
    if (element.current === null || active) {
      const params = path.decode(match.params);
      element.current = Component(params);
    }
  }

  useEffect(() => {
    if (active) {
      setMount(true);
    } else {
      const timer = setTimeout(setMount, unmount, false);
      return () => clearTimeout(timer);
    }
  }, [active, unmount, setMount]);

  const content = active || mount ? element.current : null;

  if (Transition) {
    return <Transition active={active}>{content}</Transition>;
  }

  return content;
};

export default Screen;
