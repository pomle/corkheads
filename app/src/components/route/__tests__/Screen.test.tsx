import React, { useRef } from "react";
import { MemoryRouter } from "react-router-dom";
import renderer, { act, ReactTestRenderer } from "react-test-renderer";
import { createPath, createCodec } from "lib/path";
import Screen from "../Screen";

const stringCodec = createCodec(
  (s: string) => s,
  (s: string) => s
);

const FakeTransition: React.FC<{ active: boolean }> = ({
  active,
  children,
}) => {
  return (
    <div className="transition">
      Active: {active ? "Yes" : "No"}
      <div className="transition-content">{children}</div>
    </div>
  );
};

const FakeContent: React.FC<{ params: any }> = ({ params }) => {
  const counter = useRef<number>(0);
  counter.current += 1;

  return (
    <div className="content">
      <div>Counter: {counter.current}</div>
      <div>Params: {JSON.stringify(params)}</div>
    </div>
  );
};

describe("Screen component", () => {
  let elementMock: jest.Mock;
  let transitionMock: jest.Mock;
  let testRenderer: ReactTestRenderer;

  const history = ["/router/123"];

  beforeEach(() => {
    jest.useFakeTimers();
    elementMock = jest.fn((params) => <FakeContent params={params} />);
    transitionMock = jest.fn(FakeTransition);
  });

  describe("when mounted on non-matching", () => {
    beforeEach(() => {
      const path = createPath("/not/router/url", {});
      testRenderer = renderer.create(
        <MemoryRouter initialEntries={history}>
          <Screen path={path}>{elementMock}</Screen>
        </MemoryRouter>
      );
    });

    it("has not mounted component", () => {
      const tree = testRenderer.toJSON();
      expect(tree).toMatchSnapshot();
    });

    it("has not called element", () => {
      expect(elementMock).toHaveBeenCalledTimes(0);
    });
  });

  describe("when mounted on non-matching with transition", () => {
    beforeEach(() => {
      const path = createPath("/not/router/url", {});
      testRenderer = renderer.create(
        <MemoryRouter initialEntries={history}>
          <Screen path={path} transition={transitionMock}>
            {elementMock}
          </Screen>
        </MemoryRouter>
      );
    });

    it("has not mounted component", () => {
      const tree = testRenderer.toJSON();
      expect(tree).toMatchSnapshot();
    });

    it("has set transition to not active", () => {
      expect(transitionMock).toHaveBeenCalledTimes(1);
      expect(transitionMock.mock.calls[0][0].active).toBe(false);
    });

    describe("then matching", () => {
      beforeEach(() => {
        const path = createPath("/router/123", {});
        testRenderer.update(
          <MemoryRouter initialEntries={history}>
            <Screen path={path} transition={transitionMock}>
              {elementMock}
            </Screen>
          </MemoryRouter>
        );
      });

      it("has mounted component", () => {
        const tree = testRenderer.toJSON();
        expect(tree).toMatchSnapshot();
      });

      it("has set transition to active", () => {
        expect(transitionMock).toHaveBeenCalledTimes(2);
        expect(transitionMock.mock.calls[1][0].active).toBe(true);
      });
    });
  });

  describe("when mounted on matching", () => {
    beforeEach(() => {
      const path = createPath("/router/:myId", { myId: stringCodec });
      testRenderer = renderer.create(
        <MemoryRouter initialEntries={history}>
          <Screen path={path}>{elementMock}</Screen>
        </MemoryRouter>
      );
    });

    it("has mounted component", () => {
      const tree = testRenderer.toJSON();
      expect(tree).toMatchSnapshot();
    });

    it("has called element once", () => {
      expect(elementMock).toHaveBeenCalledTimes(1);
    });

    it("has called element with params", () => {
      expect(elementMock).toHaveBeenCalledWith({ myId: "123" });
    });

    describe("then updating", () => {
      beforeEach(() => {
        const path = createPath("/router/:myId", { myId: stringCodec });
        testRenderer.update(
          <MemoryRouter initialEntries={history}>
            <Screen path={path}>{elementMock}</Screen>
          </MemoryRouter>
        );
      });

      it("presents updated mounted component", () => {
        const tree = testRenderer.toJSON();
        expect(tree).toMatchSnapshot();
      });

      it("is updating component", () => {
        expect(elementMock).toHaveBeenCalledTimes(2);
      });
    });

    describe("then unmatching", () => {
      beforeEach(() => {
        const path = createPath("/no/match", {});
        testRenderer.update(
          <MemoryRouter initialEntries={history}>
            <Screen path={path}>{elementMock}</Screen>
          </MemoryRouter>
        );
      });

      it("is still mounted", () => {
        const tree = testRenderer.toJSON();
        expect(tree).toMatchSnapshot();
      });

      it("is not updating component", () => {
        expect(elementMock).toHaveBeenCalledTimes(1);
      });

      describe("after some time", () => {
        beforeEach(() => {
          act(() => {
            jest.advanceTimersByTime(5000);
          });
        });

        // Test works, bug with tools (shrug)
        it.skip("is unmounted", () => {
          const tree = testRenderer.toJSON();
          expect(tree).toMatchSnapshot();
        });
      });
    });
  });

  describe("when mounted on over-matching", () => {
    beforeEach(() => {
      const path = createPath("/router", {});
      testRenderer = renderer.create(
        <MemoryRouter initialEntries={history}>
          <Screen path={path}>{elementMock}</Screen>
        </MemoryRouter>
      );
    });

    it("has mounted component", () => {
      const tree = testRenderer.toJSON();
      expect(tree).toMatchSnapshot();
    });

    it("has called component", () => {
      expect(elementMock).toHaveBeenCalledTimes(1);
    });

    describe("then setting exact", () => {
      beforeEach(() => {
        const path = createPath("/router", {});
        testRenderer.update(
          <MemoryRouter initialEntries={history}>
            <Screen exact path={path}>
              {elementMock}
            </Screen>
          </MemoryRouter>
        );
      });

      it("is still mounted", () => {
        const tree = testRenderer.toJSON();
        expect(tree).toMatchSnapshot();
      });

      it("has not called component", () => {
        expect(elementMock).toHaveBeenCalledTimes(1);
      });
    });
  });
});
