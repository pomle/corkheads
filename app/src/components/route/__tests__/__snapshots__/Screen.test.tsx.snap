// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Screen component when mounted on matching has mounted component 1`] = `
<div
  className="content"
>
  <div>
    Counter: 
    1
  </div>
  <div>
    Params: 
    {"myId":"123"}
  </div>
</div>
`;

exports[`Screen component when mounted on matching then unmatching after some time is unmounted 1`] = `null`;

exports[`Screen component when mounted on matching then unmatching is still mounted 1`] = `
<div
  className="content"
>
  <div>
    Counter: 
    1
  </div>
  <div>
    Params: 
    {"myId":"123"}
  </div>
</div>
`;

exports[`Screen component when mounted on matching then updating presents updated mounted component 1`] = `
<div
  className="content"
>
  <div>
    Counter: 
    2
  </div>
  <div>
    Params: 
    {"myId":"123"}
  </div>
</div>
`;

exports[`Screen component when mounted on non-matching has not mounted component 1`] = `null`;

exports[`Screen component when mounted on non-matching with transition has not mounted component 1`] = `
<div
  className="transition"
>
  Active: 
  No
  <div
    className="transition-content"
  />
</div>
`;

exports[`Screen component when mounted on non-matching with transition then matching has mounted component 1`] = `
<div
  className="transition"
>
  Active: 
  Yes
  <div
    className="transition-content"
  >
    <div
      className="content"
    >
      <div>
        Counter: 
        1
      </div>
      <div>
        Params: 
        {}
      </div>
    </div>
  </div>
</div>
`;

exports[`Screen component when mounted on over-matching has mounted component 1`] = `
<div
  className="content"
>
  <div>
    Counter: 
    1
  </div>
  <div>
    Params: 
    {}
  </div>
</div>
`;

exports[`Screen component when mounted on over-matching then setting exact is still mounted 1`] = `
<div
  className="content"
>
  <div>
    Counter: 
    1
  </div>
  <div>
    Params: 
    {}
  </div>
</div>
`;
